File: scripts/algorithms/F/First Day Where You Have Been in All the Rooms/First Day Where You Have Been in All the Rooms.py
class Solution:
    def firstDayBeenInAllRooms(self, nextVisit: List[int]) -> int:
	        n:int = len(nextVisit)
        days:List[int] = n*[0] # days[i] is the number of days it takes to first reach room i
        MOD = pow(10, 9) + 7
        for i in range(0,n-1):
            # First we are already at room i. days[i] days has passed
            days[i+1] = days[i]
            # lets go from room i to room i+1. 
            # When we first visit room i, we need to visit room i again (so room i is visited twice, which is an even number), then we can visit room i+1
            # after we fist visit room i, the next day we will visit room (nextVisit[i]). 
            days[i+1] = (days[i+1] + 1) % MOD
            # Then the problem becomes "how to go from room (nextVisit[i]) back to room i". The step is (days[i] - days[nextVisit[i]])
            days[i+1] = (days[i+1] + days[i] - days[nextVisit[i]]) % MOD
            # Then, in the next day we go from room i to i+1
            days[i+1] = (days[i+1] + 1) % MOD
        
        return days[n-1]

File: scripts/algorithms/F/First Day Where You Have Been in All the Rooms/First Day Where You Have Been in All the Rooms.cpp
#define ll long long 
int  mod=1e9+7;
class Solution {
public:
    int firstDayBeenInAllRooms(vector<int>& nextVisit) {
          int n=nextVisit.size();
          vector<vector<ll>>dp(n,vector<ll>(2,0));
          dp[0][0]=2;
          dp[0][1]=1;
          for(int i=1;i<n;i++)
          {   dp[i][1]=dp[i-1][0]+1;
               if(i==n-1)break;
              dp[i][0]=(dp[i][1]+(dp[i-1][0]-dp[nextVisit[i]][1])+2)%mod;  
          }
          return (dp[n-1][1]+mod-1)%mod;
    }
};